### PR TITLE
[scudo] Add ConditionVariable in SizeClassAllocator64

### PR DESCRIPTION
**Pull request by ChiaHungDuan**
This may improve the waiting of `Region->MMLock` while trying to refill the freelist. Instead of always waiting on the completion of `populateFreeListAndPopBatch()` or `releaseToOSMaybe()`, `pushBlocks()` also refills the freelist. This increases the chance of earlier return from `popBatches()`.

The support of condition variable hasn't been done for all platforms. Therefore, add another `popBatchWithCV()` and it can be configured in the allocator configuration by setting `Primary::UseConditionVariable` and the desired `ConditionVariableT`.

Reviewed By: cferris

Differential Revision: https://reviews.llvm.org/D156146

### CODE REVIEW

**Comment by ChiaHungDuan** [comment]
This is copied from https://reviews.llvm.org/D156146 and rebased

**Comment by ChiaHungDuan** [comment]
It has the last comment which is waiting for Hans input

File: compiler-rt/lib/scudo/standalone/condition_variable_linux.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+#include <limits.h>
+#include <linux/futex.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+
+namespace scudo {
+
+void ConditionVariableLinux::notifyAllImpl(UNUSED HybridMutex &M) {
+  const u32 V = atomic_load_relaxed(&Counter) + 1;
```
Comment (hboehm) [inline compiler-rt/lib/scudo/standalone/condition_variable_linux.cpp:25]: I withdraw my earlier comment about repeated WAKE calls. I had misread the code. I don't see any correctness issues.

To me, the code would be a bit clearer if the above line were (modulo style rules):

const u32 PrevCounter = atomic_load_relaxed(&Counter);

without the +1, but all uses of V below adjusted to make the code correct again. The test would just be LastNotifyAll != PrevCounter.

For consistency, I would do the same in waitImpl.

Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/condition_variable_linux.cpp]: Nice suggestion! Thanks

**Comment by ChiaHungDuan** [comment]
Suggested by Hans, also updated the test to only call notifyAll() after the counter increment so that we will catch the case which loses wake up.

File: compiler-rt/lib/scudo/standalone/condition_variable_base.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
+//===-- condition_variable_base.h ------------------------------------*- C++
+//-*-===//
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/condition_variable_base.h:2]: Looks like this got split across lines on accident.

**Review by cferris1000 — COMMENTED** [review]
Just the one super minor nit.

File: compiler-rt/lib/scudo/standalone/condition_variable_base.h (hunk 2 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/condition_variable_base.h]: Nice Catch!

**Review by cferris1000 — APPROVED** [review]

**Pull request merged**